File: src/roar_net_api/api/operations/random_moves_without_replacement.py
# SPDX-FileCopyrightText: © 2025 Authors of the roar-net-api-py project <https://github.com/roar-net/roar-net-api-py/blob/main/AUTHORS>
#
# SPDX-License-Identifier: Apache-2.0

from typing import Protocol, TypeVar
from collections.abc import Iterable

Solution = TypeVar("Solution", contravariant=True)
Move = TypeVar("Move", covariant=True)


class SupportsRandomMovesWithoutReplacement(Protocol[Solution, Move]):
    def random_moves_without_replacement(self, solution: Solution) -> Iterable[Move]: ...

File: julia/src/operations/random_moves_without_replacement.jl

# SPDX-FileCopyrightText: © 2025 Authors of the roar-net-api-py project <https://github.com/roar-net/roar-net-api-py/blob/main/AUTHORS>
#
# SPDX-License-Identifier: Apache-2.0
"""
    random_moves_without_replacement(neighbourhood::Neighbourhood, solution::Solution) -> iterable::Any

Return an iterable of `Move`s from `neighbourhood` for `solution` in random
order without replacement. The return type is declared `Any` because
concrete implementations may use either a `Vector{<:Move}` or a
`Channel{<:Move}`.
"""
function random_moves_without_replacement(::Neighbourhood, ::Solution)::Any end
